soccer/planning/InterpolatedPath.cpp

```diff
@@ -366,11 +365,111 @@ bool Planning::InterpolatedPath::evaluate(float t, Geometry2d::Point &targetPosO
 	return true;
 }
 
+size_t Planning::InterpolatedPath::size() const
+{
+	 return points.size();
+}
+
+bool Planning::InterpolatedPath::valid() const
+{
+	return !points.empty();
+}
+
 float Planning::InterpolatedPath::getTime(int index) const
 {
-	if (maxSpeed == -1 || maxAcceleration == -1) {
-        throw std::runtime_error("You must set maxSpeed and maxAcceleration before calling Path.evaluate()");
-    }
+	return times[index];
+}
+
 
-	return Trapezoidal::getTime(length(0,index), length(), maxSpeed, maxAcceleration, startSpeed, endSpeed);
+float Planning::InterpolatedPath::getDuration() const
+{
+	if (times.size()>0) {
+		return times.back();
+	} else {
+		return 0;
+	}
 }
+
+unique_ptr<Path> Planning::InterpolatedPath::subPath(float startTime, float endTime) const
+{
+	if (startTime<0) {
+		throw invalid_argument("startTime can't be less than zero");
+	}
+
+	if (endTime<0) {
+		throw invalid_argument("endTime can't be less than zero");
+	}
+
+	if (startTime > endTime) {
+		throw invalid_argument("startTime can't be after endTime");
+	}
+
+	InterpolatedPath *path = new InterpolatedPath();
+
+	if (startTime >= getDuration()) {
+		debugThrow(invalid_argument("startTime can't be greater than the duration of the path"));
+		return unique_ptr<Path>(path);
+	}
+
+	if (startTime == 0 && endTime>=getDuration()) {
+		return this->clone();
+	}
+
+	endTime = min(endTime, getDuration());
+
+	size_t start = 0;
+	while(times[start]<=startTime) {
+		start++;
+	}
+	start--;
+
+
+	path->times.push_back(0);
+	if (times[start]==startTime) {
+		path->points.push_back(points[start]);
+		path->vels.push_back(vels[start]);
+	} else {
+		float deltaT = (times[start+1] - times[start]);
+		float constant = (times[start+1]-startTime)/deltaT;
+		Point startPos = points[start+1]*(1-constant) + points[start]*(constant);
+		Point vi = vels[start+1]*(1-constant) + vels[start]*(constant);
+		path->points.push_back(startPos);
+		path->vels.push_back(vi);
+	}
+	Point vf;
+	Point endPos;
+	size_t end;
+	if (endTime>= getDuration()) {
```

Comment:
This `if` is kinda redundant since you do `endTime = min(endTime, getDuration());` above :)